Setup.hs

```diff
@@ -96,38 +97,65 @@ generateInterfaces pd lbi = do
   -- current directory root of the package.
 
   putStrLn "Generating Agda library interface files..."
-  forM_ (dataFiles pd) $ \fp -> when (takeExtension fp == ".agda") $ do
-    let fullpath  = ddir </> fp
-    let fullpathi = toIFile fullpath
 
-    -- remove existing interface file
-    let handleExists e | isDoesNotExistError e = return ()
+  -- The Agda.Primitive* and Agda.Builtin* modules.
+  let builtins = filter ((== ".agda") . takeExtension) (dataFiles pd)
+
+  -- Remove all existing .agdai files.
+  forM_ builtins $ \fp -> do
+    let fullpathi = toIFile (ddir </> fp)
+
+        handleExists e | isDoesNotExistError e = return ()
                        | otherwise             = throwIO e
 
     removeFile fullpathi `catch` handleExists
 
-    putStrLn $ "... " ++ fullpath
-    ok <- rawSystem' ddir agda [ "--no-libraries", "--local-interfaces"
-                               , "--ignore-all-interfaces"
-                               , "-Werror"
-                               , fullpath, "-v0"
-                               ]
-    case ok of
-      ExitSuccess   -> return ()
-      ExitFailure _ -> die $ "Error: Failed to typecheck " ++ fullpath ++ "!"
+  -- Type-check all builtin modules (in a single Agda session to take
+  -- advantage of caching).
+  let loadBuiltinCmds = concat
+        [ [ cmd ("Cmd_load " ++ f ++ " []")
+          , cmd "Cmd_no_metas"
+            -- Fail if any meta-variable is unsolved.
+          ]
+        | b <- builtins
+        , let f     = show (ddir </> b)
+              cmd c = "IOTCM " ++ f ++ " None Indirect (" ++ c ++ ")"
+        ]
+  env <- getEnvironment
+  bracket
+    (createProcess
+      ((proc agda
+          [ "--interaction"
+          , "--interaction-exit-on-error"
+          , "--no-libraries"
+          , "--local-interfaces"
+          , "-Werror"
+          , "-v0"
+          ])
+        { std_in        = CreatePipe
+        , std_out       = CreatePipe
```

Comment:
Is this correct?  Considering that under Windows, `Setup` freezes now, and the easiest explanation would be a blocked pipe, maybe you fell into a similar trap as I did, see
- https://github.com/haskell/process/issues/218

Reply:
When I put here `std_out = UseHandle stdout`, then `Setup` runs correctly!

But then, Agda prints annoying emacs commands.  So, the correct fix is probably to redirect the stdout but then swallow it in case Agda finished successfully.